tracer/src/Datadog.Trace/Debugger/Helpers/AsyncHelper.cs

```diff
@@ -0,0 +1,231 @@
+    {
+        private const string StateMachineNamePrefix = "<";
+        private const string StateMachineNameSuffix = ">";
+        private const string StateMachineFieldsPrefix = "<>";
+        private const string BuilderFieldName = StateMachineFieldsPrefix + "t__builder";
+        private const string StateFieldName = StateMachineFieldsPrefix + "1__state";
+        private const string StateMachineThisFieldName = StateMachineFieldsPrefix + "4__this";
+        private const string StateMachineLiveDebuggerAddedField = "<>dd_liveDebugger_isReEntryToMoveNext";
+        private const BindingFlags AllFieldsBindingFlags = BindingFlags.Instance | BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.FlattenHierarchy;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static FieldNameValueType[] GetHoistedArgumentsFromStateMachine<TTarget>(TTarget instance, ParameterInfo[] originalMethodParameters)
+        {
+            if (originalMethodParameters is null || originalMethodParameters.Length == 0)
+            {
+                return Array.Empty<FieldNameValueType>();
+            }
+
+            var foundFields = new FieldNameValueType[originalMethodParameters.Length];
+            var allFields = instance.GetType().GetFields(AllFieldsBindingFlags);
+
+            int found = 0;
+            for (int i = 0; i < allFields.Length; i++)
+            {
+                // we have more fields than parameters, so we break once we found all parameters
+                if (found == foundFields.Length)
+                {
+                    break;
+                }
+
+                var field = allFields[i];
+                // early exit here based on generated name pattern https://datadoghq.atlassian.net/browse/DEBUG-928
+                if (field.Name.StartsWith(StateMachineNamePrefix, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                for (var j = 0; j < originalMethodParameters.Length; j++)
+                {
+                    if (field.Name == originalMethodParameters[j].Name)
+                    {
+                        foundFields[j] = new FieldNameValueType(field.Name, field.GetValue(instance), field.FieldType);
+                        found++;
+                        break;
+                    }
+                }
+            }
+
+            return foundFields;
+        }
+
+        /// <summary>
+        /// MethodMetadataInfo saves locals from MoveNext localVarSig,
+        /// this isn't enough in async scenario because we need to extract more locals the may hoisted in the builder object
+        /// and we need to subtract some locals that exist in the localVarSig but they are not belongs to the kickoff method
+        /// For know we capturing here all locals the are hoisted (except known generated locals)
+        /// and we capturing in LogLocal the locals form localVarSig
```

Comment:
```suggestion
        /// MethodMetadataInfo saves locals from MoveNext's localVarSig,
        /// This isn't enough in the async scenario because we need to extract more locals that may have been hoisted in the builder object,
        /// and we need to subtract some locals that exist in the localVarSig but they do not 
belongs to the kickoff method
        /// For now, we're capturing all locals that are hoisted (except known generated locals),
        /// and we're capturing all the locals that appear in localVarSig with ` LogLocal `
```

Reply:
Fixed in 871a652d9b150d63443d87a542e34d4643145adf